### PR TITLE
Remove exit error types

### LINKED ISSUE #296: Consider removing exported error types
The tfexec exported error types - for example, [`tfexec.ErrCLIUsage`](https://pkg.go.dev/github.com/hashicorp/terraform-exec/tfexec#ErrCLIUsage) and [`tfexec.ErrTFVersionMismatch`](https://github.com/hashicorp/terraform-exec/blob/v0.16.1/tfexec/exit_errors.go#L279) - are implemented by parsing Terraform's human-readable error output using regular expressions.

Since this log output is subject to frequent changes and not covered by the [v1.0 compatibility promises](https://www.terraform.io/language/v1-compatibility-promises), users of these error types are at risk of having their code broken by new Terraform versions, requiring them (in the best case) to keep terraform-exec updated.

This error parsing code also seems to be a leading cause of terraform-exec nightly test failures.

Are these error types useful enough to consumers to continue maintaining them? Should we remove them prior to v1.0.0? Are there any really good use cases?

### PR DESCRIPTION
**Pull request by kmoe**
Closes #296.

This PR proposes a breaking change to the terraform-exec API removing the following error types:
`ErrConfigInvalid`
`ErrLockIdInvalid`
`ErrMissingVar`
`ErrNoConfig`
`ErrNoInit`
`ErrNoWorkspace`
`ErrStateLocked`
`ErrStatePlanRead`
`ErrTFVersionMismatch`
`ErrWorkspaceExists`

The above error types are returned based on parsing the human-readable output of Terraform CLI via regex. The human-readable output of Terraform CLI is not designed for machine parsing, and can change at any time, as it is not covered by the [Terraform v1.x Compatibility Promises](https://developer.hashicorp.com/terraform/language/v1-compatibility-promises). These changes, along with trivial text formatting issues, are a common cause of test failures in terraform-exec.

I don't think maintaining these is worth the work, in the absence of a clear use case for these errors. If you have one, please comment here or on the linked issue.  

Note that the following errors are not removed:
`ErrVersionMismatch`
`ErrNoSuitableBinary`
`ErrManualEnvVar`

In particular, the compatibility error `ErrVersionMismatch`, which is based on terraform-exec's knowledge of which flags and commands can be used with which version of Terraform, is not removed.

---

As Terraform's machine-readable UI (`-json` streaming format) matures, I expect Terraform itself is the best place to build more structure into errors and diagnostics if necessary. For example, diagnostics could contain an error symbol in addition to their human-readable summary and description.

### CODE REVIEW

File: tfexec/internal/e2etest/show_test.go [hunk 1]
```diff
@@ -111,16 +111,15 @@ func TestShow_emptyDir(t *testing.T) {
...
 	// In the basic case, in which the local backend is implicit and there are
 	// no providers to download, this is unintended behaviour, as
 	// init is not actually necessary. This is considered a known issue in
 	// pre-1.2.0 versions.
 	runTestVersions(t, []string{testutil.Latest012, testutil.Latest013, testutil.Latest014, testutil.Latest015, testutil.Latest_v1, testutil.Latest_v1_1}, "basic", func(t *testing.T, tfv *version.Version, tf *tfexec.Terraform) {
-		var noInit *tfexec.ErrNoInit
 		_, err := tf.Show(context.Background())
-		if !errors.As(err, &noInit) {
-			t.Fatalf("expected error ErrNoInit, got %T: %s", err, err)
+		if err == nil {
```
Comment (bflad) [inline tfexec/internal/e2etest/show_test.go:121]: Nit: Not sure if its worth checking for the (new) expected error in these cases or not.

**Review by bflad — APPROVED** [review]
Looks good to me 🚀

**Review by radeksimko — COMMENTED** [review]
The proposed change in principle LGTM, but there's now a number of tests which are failing, so I think we either need to reconsider the scope, or change the tests.

**Comment by kmoe** [comment]
Rebased, but this change unintentionally broke the `context.DeadlineExceeded` wrapping introduced in https://github.com/hashicorp/terraform-exec/commit/592bcd1b956ffb887a3147a22f72d2cb3dd3dfb1 (see https://github.com/golang/go/issues/21880). Fixing.

**Review by radeksimko — APPROVED** [review]
I didn't do any "deeper" checks in terms of reading through the relevant LS code, but I can confirm that all LS tests are still passing with these changes, i.e. we aren't relying on any of the error types being removed in LS.

Let's ship it!

Time will tell how much we're affected by [Hyrum's Law](https://www.hyrumslaw.com) 😁

**Comment by sergiught** [comment]
This change unfortunately makes all our test cases that expect an error fail within https://github.com/auth0/terraform-provider-auth0 e.g.

```
func TestAccBranding_WithNoCustomDomainsSet(t *testing.T) {
	acctest.Test(t, resource.TestCase{
		Steps: []resource.TestStep{
			{
				Config: testAccTenantDisallowsUniversalLoginCustomization,
				ExpectError: regexp.MustCompile(
					"managing the universal login body through the 'auth0_branding' resource " +
						"requires at least one custom domain to be configured for the tenant",
				),
			},
		},
	})
}
```

To reproduce:

```
git clone git@github.com:auth0/terraform-provider-auth0.git
cd terraform-provider-auth0
go mod vendor
make test-acc FILTER="TestAccBranding_WithNoCustomDomainsSet" // test passing
go get github.com/hashicorp/terraform-exec@v0.18.0
go mod vendor
make test-acc FILTER="TestAccBranding_WithNoCustomDomainsSet" // test failing
```

**Comment by kmoe** [comment]
Thanks @sergiught, we're looking into this now.

**Pull request merged**